### PR TITLE
fix(dot/sync): fix creating block response, fixes node sync between gossamer nodes

### PR DESCRIPTION
**Pull request by noot**
## Changes

- previously nodes were failing to create a block response because it was using the `blocktree.SubChain` function to determine the blocks to add to the response, but when the blocktree gets pruned on finalization this caused the node not to find the blocks
- update `CreateBlockResponse` to search the blockState by number instead
- also handle ascending/descending direction in the request

## Tests

```
go test ./dot/sync -short
```

## Issues

- related to #1479

### CODE REVIEW

File: dot/sync/message.go [hunk 1]
```diff
@@ -25,122 +25,154 @@ import (
...
-	}
-
-	logger.Trace("subchain", "start", subchain[0], "end", subchain[len(subchain)-1])
-
 	responseData := []*types.BlockData{}
 
-	// TODO: check ascending vs descending direction
-	for _, hash := range subchain {
+	// ascending (ie child to parent)
+	if blockRequest.Direction == 0 {
```
Comment (timwu20) [inline dot/sync/message.go:103]:
```suggestion
	switch blockRequest.Direction {
	    case 0:
	    case 1:
	}
```
nit: I like switch statements when handling finite cases.

File: dot/sync/message.go [hunk 2]
```diff
@@ -25,122 +25,154 @@ import (
...
+		return nil, err
+	}
 
-		// header
-		if (blockRequest.RequestedData & network.RequestedDataHeader) == 1 {
-			retData, err := s.blockState.GetHeader(hash)
-			if err == nil && retData != nil {
-				blockData.Header = retData.AsOptional()
-			}
+	blockData := new(types.BlockData)
```
Comment (timwu20) [inline dot/sync/message.go:134]:
```suggestion
	blockData := types.BlockData{
	   Hash: hash,
	   ....
	}
```

File: dot/sync/message.go [hunk 3]
```diff
@@ -25,122 +25,154 @@ import (
...
+	blockData.Hash = hash
+	blockData.Header = optional.NewHeader(false, nil)
+	blockData.Body = optional.NewBody(false, nil)
+	blockData.Receipt = optional.NewBytes(false, nil)
+	blockData.MessageQueue = optional.NewBytes(false, nil)
+	blockData.Justification = optional.NewBytes(false, nil)
+
+	if (requestedData & network.RequestedDataHeader) == 1 {
```
Comment (timwu20) [inline dot/sync/message.go:142]: Can you just return if `requestData == []byte(0)`

Comment (noot) [reply dot/sync/message.go:142]: yeah, will update

**Review by edwardmack — APPROVED** [review]

**Review by timwu20 — APPROVED** [review]

**Pull request merged**